src/can_protocol.py

```diff
@@ -0,0 +1,335 @@
+                    "IntegrityError: ", e)
+            finally:
+                session.close()
+
+    except json.JSONDecodeError as e:
+        customLogger.error(
+            "Error decoding JSON message: ", e)
+    except Exception as e:
+        customLogger.error(
+            "An error occurred: ", e)
+
+
+def get_current_protocol_ids():
+    """
+    Returns protocol identifiers which are sent to cloud on app startup.
+
+    Returns
+    -------
+    ids : list
+        List of protocol identifiers.
+    """
+    session = connect_to_database()
+    try:
+        # Construct a select statement to get only the `id` column
+        stmt = select(ProtocolEntity.id)
+        result = session.execute(stmt)
+        # Fetch all IDs
+        ids = [row[0] for row in result.fetchall()]
+        return ids
+    finally:
+        session.close()
+
+
+def get_data_by_can_id(can_id):
+    """
+    Function to fetch data from protocol_data_entity table based on the received can_id.
+    Used from can service module.
+
+    Parameters
+    ----------
+    can_id : int
+        CAN identifier.
+
+    Returns
+    -------
+    results : list
+        List of ProtocolDataEntity objects which have a certain CAN ID.
+    """
+    session = connect_to_database()
+    try:
+        results = session.query(ProtocolDataEntity).filter_by(can_id=can_id).all()
+    finally:
+        session.close()
+
+    return results
+
+
+def start_protocol_client(config, main_execution_flag):
+    """
+    Start Protocol MQTT subscriber client which receives data about protocol assignment/removal from the cloud.
+
+    Parameters
+    ----------
+    main_execution_flag : Event
+        Indicator for termination request for main loop.
+    config : Config
+        Enables reading parameters from config file.
+    """
+    client = mqtt_util.gcb_init_subscriber(
+        "protocol-client-id",
+        config.gateway_cloud_broker_iot_username,
+        config.gateway_cloud_broker_iot_password)
+    client.connect(config.gateway_cloud_broker_address, config.gateway_cloud_broker_port, keepalive=60)
+    mqtt_util.gcb_on_topic_subscribe(client, GCB_PROTOCOL_TOPIC)
+    # Flag passed from app module, so MQTT client doesn't run infinitely on app shutdown.
+    while not main_execution_flag.is_set():
+        client.loop(0.1)
+    client.disconnect()
+
+
+def start_protocol_startup_client(config):
+    """
+    Start Protocol MQTT startup publisher client which sends MQTT request to the cloud to get updated protocol data.
+
+    Parameters
+    ----------
+    config : Config
+        Enables reading parameters from config file.
+    """
+    client = mqtt_util.gcb_init_publisher("startup-protocol-client-id",
+                                          config.gateway_cloud_broker_iot_username,
+                                          config.gateway_cloud_broker_iot_password)
+
+    protocol_ids = get_current_protocol_ids()
+    protocol_ids_formatted = {
```

Comment:
No need to send IDs, change this to send a basic request that will trigger sync mechanism on cloud.